cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/datastore/BasicDatastore.java

```diff
@@ -934,138 +987,137 @@ public String call(SQLDatabase db) throws Exception {
+                                        String id = key[0];
+                                        String rev = key[1];
+                                        try {
+                                            BasicDocumentRevision doc = getDocumentInQueue(db, id, rev);
+                                            if (doc != null) {
+                                                AttachmentManager.addAttachmentsToRevision(db,
+                                                        attachmentsDir, doc, item.preparedAttachments.get(key));
+                                            }
+                                        } catch (DocumentNotFoundException e) {
+                                            //safe to continue, previously getDocumentInQueue could return
+                                            // null and this was deemed safe and expected behaviour
+                                            // DocumentNotFoundException is thrown instead of returning
+                                            // null now.
+                                            continue;
                                         }
-                                    } catch (DocumentNotFoundException e){
-                                        //safe to continue, previously getDocumentInQueue could return
-                                        // null and this was deemed safe and expected behaviour
-                                        // DocumentNotFoundException is thrown instead of returning
-                                        // null now.
-                                        continue;
                                     }
                                 }
+                            } catch (Exception e) {
+                                logger.log(Level.SEVERE, "There was a problem adding an " +
+                                        "attachment to the datastore", e);
+                                throw e;
                             }
-                        } catch (Exception e) {
-                            logger.log(Level.SEVERE, "There was a problem adding an " +
-                                    "attachment to the datastore", e);
-                            throw e;
-                        }
 
-                    }
-                    if (ok) {
-                        logger.log(Level.FINER, "Inserted revision: %s", rev);
-                        if (documentCreated != null) {
-                            return documentCreated;
-                        } else if (documentUpdated != null) {
-                            return documentUpdated;
+                        }
+                        if (ok) {
+                            logger.log(Level.FINER, "Inserted revision: %s", item.rev);
+                            if (documentCreated != null) {
+                                eventBus.post(documentCreated);
```

Comment:
I don't think we should post event's here. Since the create/update hasn't been officially finished yet, and in theory the whole thing can be reverted if `eventBus.post` throws an runtime exception. I also wonder what the effect this will have on user's code, because they could potentially inadvertently slow replication if they do some kind of work on the thread that calls it, which would block the db queue.